File: src/ros2_bridge/mappers/builtin_interfaces/time.rs
//! Typed mapper for `builtin_interfaces/msg/Time`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn time_to_bus(
    msg: ros_env::builtin_interfaces::msg::Time,
) -> crate::builtin_interfaces::msg::v1::Time {
    crate::builtin_interfaces::msg::v1::Time {
        sec: msg.sec.into(),
        nanosec: msg.nanosec.into(),
    }
}

pub(crate) fn time_to_ros(
    bus: crate::builtin_interfaces::msg::v1::Time,
) -> ros_env::builtin_interfaces::msg::Time {
    ros_env::builtin_interfaces::msg::Time {
        sec: bus.sec as _,
        nanosec: bus.nanosec as _,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct BuiltinInterfacesTimeMapper;

impl TypedTopicMapper for BuiltinInterfacesTimeMapper {
    type Ros = ros_env::builtin_interfaces::msg::Time;
    type Bus = crate::builtin_interfaces::msg::v1::Time;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(time_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(time_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/builtin_interfaces/time.py
"""Generated mapper for `builtin_interfaces/msg/Time`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.builtin_interfaces.msg.v1 import Time as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def time_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.sec = msg.sec
    bus.nanosec = msg.nanosec
    return bus


def time_to_ros(bus):
    from builtin_interfaces.msg import Time as RosMsg

    out = RosMsg()
    out.sec = bus.sec
    out.nanosec = bus.nanosec
    return out


class BuiltinInterfacesTimeMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from builtin_interfaces.msg import Time as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return time_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return time_to_ros(bus)
